### ISSUE TITLE
LibC GTHREADS missing.

### ISSUE THREAD

**Issue by disks86** [SerenityOS/serenity#2485]
From the looks of it gthr-default.h doesn't tie into the PThreads implementation and __GTHREADS is not defined as 1. As a result _GLIBCXX_HAS_GTHREADS is not defined. That causes std::mutex and std::condition_variable to not be available.

**Comment by ADKaster** [comment]
Serenity to avoid as much of the STL and the C++ standard library as possible on purpose.  AK/ works great for all the template and container needs, and LibPThread/LibThread seems to be doing the trick for a threading API (for now!).

At least for "native" serenity code it doesn't seem necessary to have <mutex> and <condition_variable> classes accessible.

Are you looking to port some C++ project to serenity that uses std::mutex and std::condtion_variable? In that case it might make sense to try and make those classes available.

Otherwise, a custom wrapper for pthread_* APIs in LibThread or LibPthread would be preferable.

**Comment by disks86** [comment]
Yes, I am trying to port some existing software. Specifically swift shader which uses those two classes internally.

**Comment by disks86** [comment]
In my case, it looks like Lock from LibThread will do what I need. I'm re-writing the code I'm working on to use that. I do think we'll need thread support in the c++ standard library if we intend to maintain ports of software however so I'm leaving this issue open.

**Comment by RealKC** [comment]
I ran into this too while trying to port stockfish, I discovered that gcc was compiled with "Thread model: single" and attempted to rebuild the toolchain with `--enable-thread=posix`(as we do have a LibPthread), but now libgcc  requires `<pthread.h>` when building it, and I'm not sure how to provide LibPthread's headers to it. This is what looks like to be the relevant section of the build logs: https://wastebin.party/ovakajageb.yaml

**Comment by RealKC** [comment]
So I've gotten some progress with this.

With this patch applied
```diff
diff --git a/Toolchain/BuildIt.sh b/Toolchain/BuildIt.sh
index c8f0cb502..7878707b2 100755
--- a/Toolchain/BuildIt.sh
+++ b/Toolchain/BuildIt.sh
@@ -237,30 +237,35 @@ pushd "$DIR/Build/$ARCH"
                                             --enable-languages=c,c++ \
                                             --enable-default-pie \
                                             --enable-lto \
+                                            --enable-threads=posix \
                                             ${TRY_USE_LOCAL_TOOLCHAIN:+"--quiet"} || exit 1
 
-        echo "XXX build gcc and libgcc"
+        echo "XXX build gcc"
         "$MAKE" -j "$MAKEJOBS" all-gcc || exit 1
         if [ "$(uname -s)" = "OpenBSD" ]; then
             ln -sf liblto_plugin.so.0.0 gcc/liblto_plugin.so
         fi
-        "$MAKE" -j "$MAKEJOBS" all-target-libgcc || exit 1
-        echo "XXX install gcc and libgcc"
-        "$MAKE" install-gcc install-target-libgcc || exit 1
+        echo "XXX install gcc"
+        "$MAKE" install-gcc || exit 1
 
-        echo "XXX serenity libc and libm headers"
-        mkdir -p "$BUILD"
+        echo "XXX serenity libc, libm, and pthread headers"
         pushd "$BUILD"
-            mkdir -p Root/usr/include/
             SRC_ROOT=$($REALPATH "$DIR"/..)
-            FILES=$(find "$SRC_ROOT"/Userland/Libraries/LibC "$SRC_ROOT"/Userland/Libraries/LibM -name '*.h' -print)
+            FILES=$(find "$SRC_ROOT"/Userland/Libraries/LibC "$SRC_ROOT"/Userland/Libraries/LibM "$SRC_ROOT"/Userland/Libraries/LibPthread -name '*.h' -print)
             for header in $FILES; do
-                target=$(echo "$header" | sed -e "s@$SRC_ROOT/Userland/Libraries/LibC@@" -e "s@$SRC_ROOT/Userland/Libraries/LibM@@")
+                target=$(echo "$header" | sed -e "s@$SRC_ROOT/Userland/Libraries/LibC@@" -e "s@$SRC_ROOT/Userland/Libraries/LibM@@" -e "s@$SRC_ROOT/Userland/Libraries/LibPthread@@")
                 $INSTALL -D "$header" "Root/usr/include/$target"
             done
             unset SRC_ROOT
         popd
 
+        echo "XXX build libgcc"
+        "$MAKE" -j "$MAKEJOBS" all-target-libgcc || exit 1
+        echo "XXX install libgcc"
+        "$MAKE" install-target-libgcc || exit 1
+
+
+
         echo "XXX build libstdc++"
         "$MAKE" -j "$MAKEJOBS" all-target-libstdc++-v3 || exit 1
         echo "XXX install libstdc++"
```

The crosscompiler now has the posix thread model
```sh
$ Toolchain/Local/i686/bin/i686-pc-serenity-g++ -v 2>&1 | grep "Thread model:"
Thread model: posix
```

But now I'm getting undefined references to abort when linking the kernel: https://wastebin.party/iyacivewej.coffeescript (this was done in a fresh clone to see if maybe I had artifacts interfering with the build)

Do we need to apply some patches to libstdc++ now?

**Comment by RealKC** [comment]
I think I've got a patch I can send now

**Comment by RealKC** [comment]
I think this can be closed due to #6343 having been merged
